### PR TITLE
ENH: special: Improvements for the incomplete beta functions.

### LINKED ISSUE #4244: betaincinv accuracy
betaincinv accuracy test  in `test_data.py` vs `ibeta_inv_data_ipp-ibeta_inv_data` needs pretty large tolerance:

```
Max |adiff|: 7.05547e-13
Max |rdiff|: 3.56954e-06
Bad results (1 out of 1210) for the following points (in output 0):
         0.0002742794749792665                    289206.03125              0.9688708782196045 =>          1.639978180247522e-56 !=          1.639984034231756e-56  (rdiff         3.5695373319886207e-06)
```

Could be useful to investigate if it's possible to improve this.

### LINKED ISSUE #4677: Unexpected behavior from scipy.special.btdtri
When the first parameter of a beta distribution is small and the second is large the isf returns screwy values. The following SO page illustrates the problem. 

http://stackoverflow.com/questions/29306538/unexpected-behavior-in-scipy-isf

### LINKED ISSUE #12796: BUG: nonmonotonicity in betaincinv, btdtri, stats.beta.ppf
betaincinv and related function have a nonmonotonicity which looks incorrect

statsmodels issue https://github.com/statsmodels/statsmodels/issues/7008 is for `scipy.stats.beta.isf`
but that uses btdri which is equivalent to betaincinv

```

count_ = np.arange(1, 20)
q, a, b = 1-alpha_2, count_ + 1, nobs - count_
q, a, b
(0.999995,
 array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
        19, 20]),
 array([99999, 99998, 99997, 99996, 99995, 99994, 99993, 99992, 99991,
        99990, 99989, 99988, 99987, 99986, 99985, 99984, 99983, 99982,
        99981]))

special.betaincinv(a, b, q)
array([0.00014976, 0.00017332, 0.03499907, 0.03633139, 0.03530332,
       0.00025256, 0.00027046, 0.00028788, 0.00030491, 0.00032161,
       0.00033802, 0.00035417, 0.00037009, 0.0003858 , 0.00040134,
       0.0004167 , 0.00043192, 0.00044699, 0.00046193])

inv = special.btdtri(a, b * 1., 1 - alpha_2)
inv
array([0.00014976, 0.00017332, 0.03499907, 0.03633139, 0.03530332,
       0.00025256, 0.00027046, 0.00028788, 0.00030491, 0.00032161,
       0.00033802, 0.00035417, 0.00037009, 0.0003858 , 0.00040134,
       0.0004167 , 0.00043192, 0.00044699, 0.00046193])

# inverse check

special.betainc(a, b, inv)
array([0.999995, 0.999995, 1.      , 1.      , 1.      , 0.999995,
       0.999995, 0.999995, 0.999995, 0.999995, 0.999995, 0.999995,
       0.999995, 0.999995, 0.999995, 0.999995, 0.999995, 0.999995,
       0.999995])

```​

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
* Use boost's ibeta as the backend for special.betainc
* Use boost's ibeta_inv as the backend for special.betaincinv
* Add special.betaincc and special.betainccinv as wrappers of boost's ibetac and ibetac_inv, resp.
* Deprecate special.btdtr and special.btdtri.
* Add new tests.
* Move all tests of betainc, betaincinv, betaincc and betainccinv into a separate class.
* Move all tests of beta and betaln into a separate class.

Closes gh-4244.
Closes gh-4677.
Closes gh-12796.

*Notes*

* The names for the new functions `betaincc` and `betainccinv` follow the pattern of the names of the existing incomplete gamma functions:

  ```
    Beta            Gamma
    ------------    -------------
    betainc         gammainc
    betaincinv      gammaincinv
    betaincc        gammaincc
    betainccinv     gammainccinv
  ```

### CODE REVIEW

**Comment by ev-br** [comment]
First off, this PR is great!

I've a general question about this and related efforts. What is the plan for the C cephes routines which new Boost routines are replacing.  Are they used at all after this PR is finished? If not, do we keep them (as dead code) or do we remove them completely?

**Comment by WarrenWeckesser** [comment]
Thanks @ev-br.  I added some notes to the opening comment.

I didn't remove the corresponding Cephes code for the incomplete beta function and its inverse that is now unused, but I guess that would be reasonable to do.  Does anyone care if I go ahead and remove the unused code?  All the code will live on in the git history, so it can always be recovered if necessary.

**Comment by josef-pkt** [comment]
How good are these boost functions?

In statsmodels,, I just switched away from using the boost function for the noncentral distributions (ncf, nct) in scipy.stats to use the original scipy.special functions.

I guess it will take several years to fix boost for some cases that worked without problems in scipy.special.

AFAIR, the incomplete beta functions are quite heavily used.

**Comment by WarrenWeckesser** [comment]
> How good are these boost functions?

An example of the goodness: this PR fixes https://github.com/scipy/scipy/issues/12796.

I also ran some tests to check the relative error of the results of the old and new versions of `btdtr` and `btdtri`.

For `btdtr` (aka `betainc`), I used the following script.  It runs `btdtr` on a grid in (a, b, x) space, and computes the relative error of the results by computing the true value with mpmath.

<details>
<summary> check_btdtr_rel_error.py </summary>

```

import numpy as np
import scipy
from scipy.special import btdtr
from mpmath import mp


def _mp_betainc(a, b, x, dps=100):
    with mp.workdps(dps):
        return float(mp.betainc(a, b, x1=0, x2=x, regularized=True))


mp_betainc = np.vectorize(_mp_betainc, otypes=[float])


def relative_error(y, ref):
    delta = np.abs(y - ref)
    with np.errstate(invalid='ignore'):
        return np.where((ref == 0) & (y == 0), 0.0, delta/ref)


n = 50
param_min = 1e-9
param_max = 2000
x_min = 1e-12
a = np.geomspace(param_min, param_max, n)[:, None, None]
b = np.geomspace(param_min, param_max, n)[:, None]
s = np.geomspace(x_min, 0.5, n)
x = np.concatenate((s, s[-2::-1]))

print(f'scipy version {scipy.__version__}')
print(f'sample size {a.size * b.size * x.size}')
print(f'a, b in [{param_min}, {param_max}]')
print(f'x in [{x_min}, {1 - x_min}]')

beta1 = btdtr(a, b, x)
beta1_mp = mp_betainc(a, b, x)

relerr = relative_error(beta1, beta1_mp)

print(f'max relative error: {relerr.max()}')
```

</details>

*Results*

Main development branch:

```
scipy version 1.11.0.dev0+1652.9da5381
sample size 247500
max relative error: 4.6696703954562166e-12
```

This PR (boost-based):

```
scipy version 1.11.0.dev0+1657.6e59333
sample size 247500
max relative error: 3.4774136976727376e-16
```

So the new code lowers the worst-case error from 4.7e-12 to 3.5e-16.

For `btdtri` (aka `betaincinv`), I used the following script.  It computes the relative error of `btdri` on a random sample from (a, b, x) space. I don't have a robust mpmath-based implementation of the inverse function.  The function `mpsci.distribution.beta.invcdf` use `mpmath`'s `findroot` function, but there are large regions of parameter space where it does not find the root.  So the script doesn't probe the extremes of the parameter space as much as the previous one does.

<details>
<summary> check_btdtri.py </summary>

```
import numpy as np
import scipy
from scipy.special import btdtri
from mpmath import mp
from mpsci.distributions import beta


def _mp_betainc(a, b, x, dps=100):
    with mp.workdps(dps):
        return float(mp.betainc(a, b, x1=0, x2=x, regularized=True))


mp_betainc = np.vectorize(_mp_betainc, otypes=[float])


def _mp_betaincinv(a, b, p, dps=100):
    # print(a, b, p)
    with mp.workdps(dps):
        return float(beta.invcdf(p, a, b))


mp_betaincinv = np.vectorize(_mp_betaincinv, otypes=[float])


def relative_error(y, ref):
    delta = np.abs(y - ref)
    with np.errstate(invalid='ignore'):
        return np.where((ref == 0) & (y == 0), 0.0, delta/ref)


rng = np.random.default_rng(1212631374725253140650)
n = 20000
a = 2*rng.gamma(3, size=n)
b = 2*rng.gamma(3, size=n)
x = np.linspace(0, 1, n+2)[1:-1]

print(f'scipy version {scipy.__version__}')
print(f'sample size {x.size}')

y = btdtri(a, b, x)
y_mp = mp_betaincinv(a, b, x)
relerr = relative_error(y, y_mp)

print(f'max relative error: {relerr.max()}')
```

</details>

*Results*

Main development branch:

```
scipy version 1.11.0.dev0+1652.9da5381
sample size 20000
max relative error: 8.100858865683033e-15
```

This PR (boost-based):

```
scipy version 1.11.0.dev0+1657.6e59333
sample size 20000
max relative error: 1.939684499123348e-16
```

The version based on the boost function lowers the worst-case relative error from 8e-15 to 2e-16.

> I guess it will take several years to fix boost for some cases that worked without problems in scipy.special.

There have been issues, mostly the result of the SciPy tests flushing out edge cases that were handled either inconsistently or incorrectly in boost.  In fact,  I ran into another edge case while working with the above scripts.  For a particular input with `x` *extremely* small, the boost function `ibeta_inv` throws a C++ exception because an assertion in the code fails.  The issue was reported in https://github.com/boostorg/math/issues/961.  I've moved this PR to draft until that issue is resolved.

This doesn't worry me.  The boost developers have been very responsive to issues coming out of our use in SciPy.  In most cases, problems have been addressed in days, or at most a couple weeks. The fixes for any issues that aren't discovered until after a SciPy release can most likely be backported in a point release.  The advantages of using boost/math, with its active and knowledgeable maintainers, make it worthwhile dealing with the kind of issues we've seen while making the transition.

**Comment by josef-pkt** [comment]
I understand the move to boost, eventually it will have much better support and wider usebase than old scipy.special code.

However, I would like to avoid the transition for myself and statsmodels users until the dust has settled.
scipy.special got 10 or more years of improvements and works for our edge cases.
I can still avoid scipy.stats distributions if I have access to scipy special. But if scipy special switches, then we cannot avoid boost anymore.

**Comment by tupui** [comment]
What is the status here? The boost issues seems fixed now. Do we want to see this in 1.11?

**Comment by WarrenWeckesser** [comment]
When I added a test (locally, not pushed to github) of `betaincinv(a, b, p)` with `a = b = 14.208308325339239` and `p = 7.703145458496392e-307` as a regression test for the boost issue, I found that while the crash is fixed, the result computed has no significant digits correct (`5.293955920339377e-23` instead of the expected `8.566004561846704e-23`).  The test is an extreme edge case: `a` and `b` are relatively large, and `p` is close to the smallest normal float64. But the old code is able to compute this result with a relative error of about 2e-15.  I'm going to leave this marked as *draft* while I investigate further.

**Comment by WarrenWeckesser** [comment]
I figured out a fix for the issue mentioned in my previous comment.  I had to disable boost/math's internal promotion of `double` to `long double` (we also do that in our use of boost/math in `scipy.stats`), and increased the default maximum number of root-finding iterations.  Without the latter, the boost/math function `ibeta_inv` would give up and report convergence failure for cases where the Cephes function `incbi` would give good results.

File: scipy/special/tests/test_basic.py
Comment (lorentzenchr) [inline scipy/special/tests/test_basic.py:1320]:
```suggestion
    #    x = mp.findroot(
    #        lambda t: mp.betainc(a, b, 0, t, regularized=True) - p,
    #        (8e-21, 9e-21),
    #        solver='anderson',
    #    )
```
Unimportant cosmetics. I find it easier to read.

**Review by lorentzenchr — APPROVED** [review]
Seems like a very solid PR to me. Thanks @WarrenWeckesser (I have to return some favor for #11313).

I there a larger roadmap / issue to replace all of cephes?

File: scipy/special/__init__.py
Comment (j-bowhay) [inline scipy/special/__init__.py:804]: Should we deprecate these rather than carry two sets of names for the same thing, either here or in a follow-up PR?

File: scipy/special/boost_special_functions.h
Comment (WarrenWeckesser) [reply scipy/special/boost_special_functions.h:58]: Nice!  I assume you meant `ibeta_wrap<float>` here.

Even with that correction, scipy fails to build with this change.  Some coordination with the ufunc code generation might be necessary.  I think this bit of code reduction can be deferred to a follow-up.

**Comment by WarrenWeckesser** [comment]
Thanks @lorentzenchr and @j-bowhay.  I'll be updating the pull request soon, and I'll add code to properly deprecate the old names and address any other comments.

Boost/math also has functions for the inverse of the incomplete beta function with respect to the parameters `a` and `b`.  It would be consistent with the names used in this PR to define `betaincinva`, `betaincinvb`, `betainccinva` and `betainccinvb` as wrappers of the corresponding boost functions `ibeta_inva`, `ibeta_invb`, `ibetac_inva` and `ibetac_invb`, resp., and deprecate `btdtria` and `btdtrib`.  I think that can be done after this PR is completed.

**Comment by h-vetinari** [comment]
> I'll be updating the pull request soon, and I'll add code to properly deprecate the old names and address any other comments.

I guess you're aware that 1.11 is set to branch soon (end of May)? Not that we won't survive if this PR doesn't land before that, but just as a reminder in case that's your target. :)

File: scipy/special/__init__.py
Comment (WarrenWeckesser) [reply scipy/special/__init__.py:804]: `bdtr` is still a wrapper of a Cephes routine. This PR doesn't make any changes to `bdtr`.

**Comment by WarrenWeckesser** [comment]
In the last update, `btdtr` and `btdtri` are deprecated but otherwise untouched from the `main` branch.

File: scipy/special/_add_newdocs.py
Comment (WarrenWeckesser) [reply scipy/special/_add_newdocs.py]: This is fixed in the latest update.

File: scipy/special/_add_newdocs.py
Comment (WarrenWeckesser) [reply scipy/special/_add_newdocs.py]: Done.

File: scipy/special/__init__.py
Comment (WarrenWeckesser) [reply scipy/special/__init__.py]: Done.

**Comment by WarrenWeckesser** [comment]
The short summary of test failure of the job "Linux Meson tests / Meson build (3.11, false) (pull_request)" is
```
FAILED scipy/sparse/csgraph/tests/test_spanning_tree.py::test_minimum_spanning_tree - AssertionError: 
Arrays are not equal
```
That is not related to this PR.

**Comment by WarrenWeckesser** [comment]
If this is merged, the commits should be squashed, and the commit message edited to
```
ENH: special: Improvements for the incomplete beta functions. #17697 

* Use boost's ibeta as the backend for special.betainc
* Use boost's ibeta_inv as the backend for special.betaincinv
* Add special.betaincc and special.betainccinv as wrappers of boost's ibetac and ibetac_inv, resp.
* Deprecate special.btdtr and special.btdtri.
* Add new tests.
* Move all tests of betainc, betaincinv, betaincc and betainccinv into a separate class.
* Move all tests of beta and betaln into a separate class.

Closes gh-12796.

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
```

File: scipy/special/__init__.py
Comment (j-bowhay) [inline scipy/special/__init__.py:827]: We should make sure that these warnings state when they will be removed

File: scipy/special/tests/test_data.py
Comment (j-bowhay) [inline scipy/special/tests/test_data.py:567]: Would prefer to use the new name rather than filter the warnings

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Couple of comments around the deprecation

File: scipy/special/_add_newdocs.py
Comment (WarrenWeckesser) [reply scipy/special/_add_newdocs.py:1342]: Forgot about this in my last update.  It will be in the next one.

File: scipy/special/tests/test_data.py
Comment (WarrenWeckesser) [reply scipy/special/tests/test_data.py:567]: In the recent update, I restored `btdtr` and `bdtri` to their original wrappers of the Cephes functions, and deprecated them.  The tests for these functions should still be run.

We *could* make the filter more specific, so the deprecation warning is only filtered for `btdtr` and `btdtri`.  I guess the point of such a change would be to avoid accidentally filtering an incorrect or unexpected `DeprecationWarning`?

File: scipy/special/__init__.py
Comment (WarrenWeckesser) [reply scipy/special/__init__.py]: I added information about SciPy versions that will show up in the docstring and in the deprecation warning.

**Comment by WarrenWeckesser** [comment]
I added a helper function for deprecating ufuncs in `scipy.special`.  The helper function `deprecate_ufunc()` modifies the beginning of the docstring so that the first line of the docstring--the line that is generated by the NumPy ufunc machinery--remains the first line of the deprecated function.  For example, with this helper function, the first line of the `btdtr` docstring remains

```
btdtr(x1, x2, x3, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])
```

If we use `np.deprecate` to deprecate `btdtr` without further modifying the docstring, that line would end up a few lines into the docstring (following the text that was added by `np.deprecate()`), and Sphinx will treat it as plain text.  The occurrence of `*` in that line will be treated as Sphinx markup, which is not what we want.  (Sphinx will generate a warning because there is no closing `*` to complete the emphasis markup.)

File: scipy/special/tests/test_data.py
Comment (h-vetinari) [reply scipy/special/tests/test_data.py:567]: This should be as easy as replacing `"btdtr"` -> `"betainc"`, `"btdtri"` -> `"betaincinv"` in the definition of `BOOST_TESTS` above. Then you can remove the warning filter again.

**Review by h-vetinari — COMMENTED** [review]
I haven't verified the correctness of the wrappers, but basically this LGTM

> I added a helper function for deprecating ufuncs in `scipy.special`.

I find this a bit overkill for two functions TBH, rather than, say, just doing:
```python
def btdtr(*args, **kwargs):
    warnings.warn(depr_msg, category=DeprecationWarning, stacklevel=2)
    return betainc(*args, **kwargs)
btdtr.__doc__ = betainc.__doc__  # possibly plus deprecation also in docstring
```
Which would be ~10 lines rather than 70.

I also agree with @j-bowhay to update the test names rather than filter the warnings, _but_ both of those issues are IMO not nearly severe enough to miss the 1.11 window over, so I'd be fine with a merge as-is (though I'd prefer at least the point below to be fixed).

File: scipy/special/tests/test_data.py
Comment (WarrenWeckesser) [reply scipy/special/tests/test_data.py]: `betainc` and `betaincinv` are already in `BOOST_TESTS`.  As I noted earlier, `btdtr` and `btdri` are no longer aliases of `betainc` and `betaincinv`; the only changes made to `btdtr` and `btdtri` are the deprecation.  Since they are using different code than that of `betainc` and `betaincinv`, their tests should still be run.  When `btdtr` and `btdtri` are removed at the end of their deprecation period, these tests can be removed.

Answering my own question above, I think it is better to only filter expected deprecation warnings, so updated the test accordingly.

**Comment by WarrenWeckesser** [comment]
>>  I added a helper function for deprecating ufuncs in scipy.special.
>
> I find this a bit overkill for two functions 

If we deprecate more ufuncs (and we will), a utility wrapper generator would reduce code duplication and help ensure consistency.  That's what I had in mind with `deprecate_ufunc()`.  But I don't particularly like how `np.deprecate` modifies the docstring, so I updated the code to use the custom wrappers suggested above.  Eventually it might be nice to have our own alternative to `np.deprecate()`.

But now `mypy` is not happy with the redefinition of `btdtr` and `btdtri` in `scipy/special/__init__.py`.  I'm looking into that, but if anyone is familiar with a problem like that, suggestions would be welcome.

**Comment by WarrenWeckesser** [comment]
The last commit "fixes" (i.e. ignores) these errors reported by `mypy`:

```
scipy/special/__init__.py:829: error: Name "btdtr" already defined (possibly by an import)  [no-redef]
scipy/special/__init__.py:834: error: Property "__doc__" defined in "ufunc" is read-only  [misc]
scipy/special/__init__.py:837: error: Name "btdtri" already defined (possibly by an import)  [no-redef]
scipy/special/__init__.py:842: error: Property "__doc__" defined in "ufunc" is read-only  [misc]
```

**Comment by WarrenWeckesser** [comment]
I think I've responded to all the comments.  I'm still filtering out the deprecation warnings for the reasons given above, but the filter is only applied to the functions that are expected to generate deprecation warnings.

**Comment by tylerjereddy** [comment]
> I think I've responded to all the comments. 

Great, if someone is confident enough to merge they can restore the milestone back if it happens before I branch, but I'm going to bump it because this shouldn't be a release blocker and I do need to branch on May 29th.

**Comment by WarrenWeckesser** [comment]
Merged main. The two CI failures are unrelated (https://github.com/scipy/scipy/issues/19036).

**Comment by dschmitz89** [comment]
Flying by as I found two other old related issues: https://github.com/scipy/scipy/issues/4677 and https://github.com/scipy/scipy/issues/4677

Would they also get closed by this PR?

**Comment by steppi** [comment]
@WarrenWeckesser I'll try to give this a proper review within 1-2 weeks. From a quick glance I don't anticipate any issues and will hopefully just be able to merge it without changes.

**Comment by WarrenWeckesser** [comment]
@dschmitz89, great find.  I just checked, and this pull request should close both those issues.  I'll update the unit tests tomorrow to include tests cases from those issues.

@steppi, thanks!

**Comment by WarrenWeckesser** [comment]
Instead of updating the tests today, I used my time machine to go back to January and add tests for #4244 and #4677 to my very first commit. :face_with_spiral_eyes: 

(Yeah, its been long enough that I forgot that I had already verified that the PR fixes those issues.)

**Comment by dschmitz89** [comment]
> Instead of updating the tests today, I used my time machine to go back to January and add tests for #4244 and #4677 to my very first commit. 😵‍💫
> 
> (Yeah, its been long enough that I forgot that I had already verified that the PR fixes those issues.)

What about adding in the PR description at the top that these issues are also closed by this PR? Otherwise they might be forgotten after merging.

**Comment by WarrenWeckesser** [comment]
> What about adding in the PR description at the top that these issues are also closed by this PR?

Done.

**Review by steppi — APPROVED** [review]
Very nice. The improvement is staggering. I created a large benchmark set similar to what I did with `hyp2f1` and there was not a single instance where one of the previous implementations had better relative error. I'll put that together into something I can share over the weekend.

**Pull request merged**
